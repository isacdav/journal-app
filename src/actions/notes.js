import Swal from 'sweetalert2';
import { db } from '../firebase/firebase';
import { fileUpload } from '../helpers/fileHelper';
import { loadNotes } from '../helpers/notesHelper';
import { types } from '../types/types';

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime(),
    };

    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activateNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  };
};

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = uid => {
  return async dispatch => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const activateNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const setNotes = notes => ({
  type: types.notesLoad,
  payload: notes,
});

export const startSaveNote = note => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) delete note.url;

    const noteToDb = { ...note };
    delete noteToDb.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToDb);

    dispatch(refreshNote(note.id, noteToDb));

    Swal.fire('Saved', note.title, 'success');
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const startUploadingFile = file => {
  return async (dispatch, getState) => {
    const { active } = getState().notes;

    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait while the picture is being uploaded.',
      allowOutsideClick: false,
      willOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    active.url = fileUrl;

    dispatch(startSaveNote(active));

    Swal.close();
  };
};

export const startDeleting = id => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    await db.doc(`${uid}/journal/notes/${id}`).delete();

    dispatch(deleteNote(id));
  };
};

export const deleteNote = id => ({
  type: types.notesDelete,
  payload: id,
});

export const cleanNotesLogout = () => ({
  type: types.notesLogoutCleaning,
});
