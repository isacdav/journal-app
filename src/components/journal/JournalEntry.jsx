import React from 'react';
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activateNote } from '../../actions/notes';

const JournalEntry = ({ id, body, title, url, date }) => {
  const noteDate = moment(date);
  const dispatch = useDispatch();

  const handleEntryClick = () => {
    dispatch(activateNote(id, { date, title, body, url }));
  };

  return (
    <div className='journal__entry' onClick={handleEntryClick}>
      {url && (
        <div
          className='journal__entry-pic'
          style={{
            backgroundSize: 'cover',
            backgroundImage: `url(${url})`,
          }}></div>
      )}
      <div className='journal__entry-body'>
        <p className='journal__entry-tittle'>{title}</p>
        <p className='journal__entry-content'>{body}</p>
      </div>
      <div className='journal__entry-date'>
        <span>{noteDate.format('ddd')}</span>
        <h4>{noteDate.format('D')}</h4>
      </div>
    </div>
  );
};

export default JournalEntry;
