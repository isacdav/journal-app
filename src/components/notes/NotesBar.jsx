import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploadingFile } from '../../actions/notes';

const NotesBar = () => {
  const dispatch = useDispatch();

  const { active } = useSelector(state => state.notes);

  const handleSave = () => {
    dispatch(startSaveNote(active));
  };

  const handlePictureClick = () => {
    document.querySelector('#file_selector').click();
  };

  const handleFileChange = e => {
    const file = e.target.files[0];
    if (file) {
      dispatch(startUploadingFile(file));
    }
  };

  const currentDate = new Date().toDateString();

  return (
    <div className='notes__bar'>
      <span>{currentDate}</span>
      <input
        id='file_selector'
        type='file'
        style={{ display: 'none' }}
        onChange={handleFileChange}
      />
      <div>
        <button className='btn' onClick={handlePictureClick}>
          Picture
        </button>
        <button className='btn' onClick={handleSave}>
          Save
        </button>
      </div>
    </div>
  );
};

export default NotesBar;
