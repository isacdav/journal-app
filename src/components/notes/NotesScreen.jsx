import React, { useRef, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activateNote, startDeleting } from '../../actions/notes';
import useForm from '../../hooks/useForm';
import NotesBar from './NotesBar';

const NotesScreen = () => {
  const dispatch = useDispatch();

  const { active: note } = useSelector(state => state.notes);
  const [formValues, handleInputChange, reset] = useForm(note);
  const { title, body, id } = formValues;

  // Save the current id to compare later
  const activeId = useRef(note.id);

  const handleDelete = () => {
    dispatch(startDeleting(id));
  };

  useEffect(() => {
    // Validate if the note has changed
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activateNote(formValues.id, { ...formValues }));
  }, [formValues, dispatch]);

  return (
    <div className='notes__main'>
      <NotesBar />
      <div className='notes__content'>
        <input
          type='text'
          name='title'
          placeholder='My title'
          className='notes__title'
          autoComplete='off'
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          name='body'
          placeholder='A description of my day...'
          className='notes__textarea'
          value={body}
          onChange={handleInputChange}></textarea>
        {note.url && (
          <div className='notes__image'>
            <img src={note.url} alt='My day photography' />
          </div>
        )}
      </div>

      <button className='btn btn-danger' onClick={handleDelete}>
        Delete
      </button>
    </div>
  );
};

export default NotesScreen;
