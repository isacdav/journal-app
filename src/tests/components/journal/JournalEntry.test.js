import '@testing-library/jest-dom';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import JournalEntry from '../../../components/journal/JournalEntry';
import { activateNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const note = {
  id: 'noteid',
  title: 'notetitle',
  body: 'notebody',
  date: 0,
  url: 'https://www.imagenes.com/qouydboi',
};

const wrapper = mount(
  <Provider store={store}>
    <JournalEntry {...note} />
  </Provider>
);

describe('<JournalEntry /> Tests', () => {
  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should set note as active', () => {
    wrapper.find('.journal__entry').prop('onClick')();

    expect(store.dispatch).toHaveBeenCalledWith(
      activateNote(note.id, { ...note })
    );
  });
});
