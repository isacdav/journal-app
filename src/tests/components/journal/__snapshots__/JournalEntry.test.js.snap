// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JournalEntry /> Tests should match snapshot 1`] = `
<div
  className="journal__entry animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-pic"
    style={
      Object {
        "backgroundImage": "url(https://www.imagenes.com/qouydboi)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-tittle"
    >
      notetitle
    </p>
    <p
      className="journal__entry-content"
    >
      notebody
    </p>
  </div>
  <div
    className="journal__entry-date"
  >
    <span>
      Wed
    </span>
    <h4>
      31
    </h4>
  </div>
</div>
`;
